### PR TITLE
[Agent] support protocol parse port range

### PR DESCRIPTION
**Pull request by qq906907952**
### This PR is for:
- Agent

#### Checklist
- [ ] Added unit test.
#### Backport to branches
- main

### CODE REVIEW

File: agent/crates/public/src/bitmap.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Error {
+    IndexOutOfBound,
+}
```
Comment (rvql) [inline agent/crates/public/src/bitmap.rs:20]: 换行

File: agent/crates/public/src/bitmap.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+#[derive(Eq, Clone)]
```
Comment (rvql) [inline agent/crates/public/src/bitmap.rs:21]: derive PartialEq

File: agent/crates/public/src/bitmap.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        self.0.len() * 8 - 1
+    }
+
+    // return vec index and u8 bit
+    fn get_idx_pos(&self, pos: usize) -> (usize, u8) {
+        (pos / 8, (pos % 8) as u8)
+    }
+
+    fn get_by_idx_pos(&self, idx: usize, u_pos: u8) -> bool {
+        !(self.0.get(idx).unwrap() & (1 << u_pos) == 0)
```
Comment (rvql) [inline agent/crates/public/src/bitmap.rs:66]: `self.0.get(idx).unwrap() & (1 << u_pos) != 0`

File: agent/src/common/l7_protocol_log.rs [hunk 4]
```diff
@@ -14,11 +14,12 @@
  * limitations under the License.
  */
 
-use std::net::IpAddr;
-
 use enum_dispatch::enum_dispatch;
+use public::bitmap::Bitmap;
 use public::enums::IpProtocol;
 use public::l7_protocol::L7Protocol;
+use std::net::IpAddr;
```
Comment (rvql) [inline agent/src/common/l7_protocol_log.rs:21]: use要排序
```
标准库
第三方
super
crate
public
```

File: agent/src/config/config.rs [hunk 5]
```diff
@@ -417,6 +421,45 @@ impl YamlConfig {
...
         Ok(())
     }
+
+    pub fn get_protocol_port_parse_bitmap(&self) -> Vec<(String, Bitmap)> {
+        /*
+            parse all protocol port range
+            format example:
+
+                l7-protocol-ports:
+                    "HTTP": ["80","8080","1000-20000"]
```
Comment (sharang) [inline agent/src/config/config.rs:431]: 确认下 server 端下发的是数组还是字符串？

File: agent/src/common/l7_protocol_log.rs [hunk 6]
```diff
@@ -281,6 +282,23 @@ all_protocol!(
     // add protocol below
 );
 
+impl L7ProtocolParser {
+    pub fn is_skip_parse_by_port_bitmap(
+        &self,
+        port_bitmap: Iter<(String, Bitmap)>,
```
Comment (rvql) [inline agent/src/common/l7_protocol_log.rs:288]: 这里参数为啥不用`&Vec<(String, Bitmap)>`

File: agent/src/config/config.rs [hunk 7]
```diff
@@ -417,6 +421,45 @@ impl YamlConfig {
...
+                ...
+        */
+        let mut port_bitmap = Vec::new();
+        for (key, val) in self.l7_protocol_ports.iter() {
+            let mut bitmap = Bitmap::new(u16::MAX as usize, false);
+
+            for s in val.iter() {
+                if let Ok(port) = s.trim().parse::<u16>() {
+                    if let Ok(_) = bitmap.set(port as usize, true) {};
```
Comment (rvql) [inline agent/src/config/config.rs:440]: 不嫌难看吗？
`let _ = bitmap.set(port as usize, true);`

File: agent/src/config/handler.rs [hunk 8]
```diff
@@ -31,6 +31,7 @@ use cgroups_rs::{CpuResources, MemoryResources, Resources};
 use flexi_logger::writers::FileLogWriter;
 use flexi_logger::{Age, Cleanup, Criterion, FileSpec, LoggerHandle, Naming};
 use log::{info, warn, Level};
+use public::bitmap::Bitmap;
```
Comment (rvql) [inline agent/src/config/handler.rs:34]: use顺序

File: agent/crates/public/src/bitmap.rs [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            return Err(Error::IndexOutOfBound);
+        }
+
+        let (idx, u_pos) = self.get_idx_pos(pos);
+        let old = self.get_by_idx_pos(idx, u_pos);
+        self.set_by_idx_pos(idx, u_pos, val);
+        Ok(old)
+    }
+
+    pub fn set_range(&mut self, start: usize, end: usize, val: bool) -> Result<(), Error> {
```
Comment (rvql) [inline agent/crates/public/src/bitmap.rs:43]: https://doc.rust-lang.org/std/ops/struct.Range.html

File: agent/crates/public/src/bitmap.rs [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        if start > end {
+            return Err(Error::RangeStartGtEnd);
+        }
+        if end > self.get_max_pos() {
+            return Err(Error::IndexOutOfBound);
+        }
+        let (start_idx, start_u_pos) = self.get_idx_pos(start);
+        let (end_idx, end_u_pos) = self.get_idx_pos(end);
+
+        for i in start_u_pos..=if end_idx > start_idx { 7 } else { end_u_pos } {
```
Comment (rvql) [inline agent/crates/public/src/bitmap.rs:53]: 逻辑不清楚
```
if end_idx == start_idx {
     xxxx
     return
}
处理开头
处理中间
处理结尾
```

**Review by yuanchaoa — APPROVED** [review]

**Review by rvql — APPROVED** [review]

**Review by yuanchaoa — APPROVED** [review]

**Review by sharang — APPROVED** [review]

**Pull request merged**